### DISCUSSION TITLE
do.call dependencies not detected?

### DISCUSSION THREAD

**Discussion by tagteam** [ropensci/targets#831] [General]
first let me say that this package really changes my world. the online material is excellent and the package is really what I needed to improve my readily quite efficient workflow. here is a quick question: I noticed that the dependence on g is not detected when do.call is used, c.f., tar_visnetwork(). here is a minimal example of my _targets.R file:

```r
g <- function(x){x}
f1 <- function(y){
    d = do.call("g",list(y))
}
f2 <- function(y){
    g(y)
}
a <- tar_target(A,{f1(1)})
b <- tar_target(B,{f2(1)})
list(a,b)
```
to circumvent this problem, is there a way to note the dependence on g? I could not see an argument of tar_target that would allow me to specify the dependence...

**Comment by wlandau** [comment]
This is an inherent limitation, the arguments to `do.call()` probably won't be detected as global symbols: https://books.ropensci.org/targets/targets.html#dependencies. Static code analysis is never perfect, so it is good that you are checking the graph. To force a dependency, you can mention `g` explicitly as a symbol in `f1()`:

```{r}
f1 <- function(y) {
  g # forces a dependency on g
  do.call("g", list(x = y))
}
```

**Reply by tagteam** [reply]
thanks a lot, this works!

**Reply by alexpghayes** [reply]
I think a quick example of manually declaring dependencies would be a great addition to the docs book!

Just came across this in a slightly different setting with formulas, included below just for giggles.

```
f <- function(x) {
  x^2
}

fit_lm <- function(data = mtcars) {
  formula <- as.formula("f(mpg) ~ hp")
  lm(formula, data = data)
}

fit_lm2 <- function(data = mtcars) {
  lm(f(mpg) ~ hp, data = data)
}

tar_deps(fit_lm)
tar_deps(fit_lm2)
tar_deps_raw(fit_lm)
tar_deps_raw(fit_lm2)
```

**Reply by wlandau** [reply]
I just rewrote the section on dependencies (https://github.com/ropensci-books/targets/commit/7b1f99100536154a41391236806318aedd04464c) and it should go online soon.

**Reply by alexpghayes** [reply]
Looks great!
